src/pldm_fw.cpp

```diff
@@ -8,32 +8,161 @@
 #include <libpldm/pldm.h>
 #include <libpldm/state_set.h>
 
+#ifdef PLDM_HAS_POLL
+#include <poll.h>
+#endif
+
 #include <iostream>
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
 
 namespace panel
 {
-types::Byte PldmFramework::getInstanceID()
+pldm_instance_id_t PldmFramework::getInstanceID()
 {
-    types::Byte instanceId = 0;
-    auto bus = sdbusplus::bus::new_default();
+    // Fetch Instance ID DB
+    int rc = pldm_instance_db_init_default(&pldmInstanceIdDb);
 
-    try
+    if (rc < 0)
     {
-        auto method = bus.new_method_call(
-            "xyz.openbmc_project.PLDM", "/xyz/openbmc_project/pldm",
-            "xyz.openbmc_project.PLDM.Requester", "GetInstanceId");
-        method.append(mctpEid);
-        auto reply = bus.call(method);
-        reply.read(instanceId);
+        throw std::runtime_error(
+            "Call to pldm_instance_db_init_default failed with return code " +
+            std::to_string(rc));
     }
-    catch (const sdbusplus::exception::SdBusError& e)
+
+    // Initialising with an invalid/unused ID.
+    pldm_instance_id_t instanceID = constants::PLDM_INST_ID_MAX + 1;
+
+    rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    if (rc == -EAGAIN)
     {
-        std::cerr << e.what() << std::endl;
-        throw FunctionFailure("pldm: call to GetInstanceId failed.");
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    }
+
+    if (rc < 0)
+    {
+        throw FunctionFailure(
+            "Call to pldm_instance_id_alloc failed with return code " +
+            std::to_string(rc));
+    }
+
+    std::cout << "Got instanceId: " << static_cast<int>(instanceID)
+              << " from PLDM eid: " << static_cast<int>(tid) << std::endl;
+    return instanceID;
+}
+
+void PldmFramework::freeInstanceID(pldm_instance_id_t instanceID)
+{
+    if (!pldmInstanceIdDb)
+    {
+        return;
+    }
+
+    // constants::PLDM_INST_ID_MAX is 32 (0 to 31)
+    if (instanceID < constants::PLDM_INST_ID_MAX)
+    {
+        int rc = pldm_instance_id_free(pldmInstanceIdDb, tid, instanceID);
+
+        if (rc == -EAGAIN)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            rc = pldm_instance_id_free(pldmInstanceIdDb, tid, instanceID);
+        }
+
+        if (rc < 0)
+        {
+            std::cerr << "pldm_instance_id_free failed to free id = "
+                      << static_cast<int>(instanceID)
+                      << " of tid = " << static_cast<int>(tid) << " rc = " << rc
+                      << std::endl;
+        }
+    }
+
+    // Destroy the instance ID DB
+    int rc = pldm_instance_db_destroy(pldmInstanceIdDb);
+    if (rc < 0)
```

Comment:
Why is this being checked for less than 0? Called API will always return 0.

Reply:
On failure libpldm apis will return value less than 0.

The below link has the requester's expected return value on success and failure. https://github.com/openbmc/libpldm/blob/5192e2e26f26ffbbf4256429ea5ae42237e011cc/include/libpldm/pldm.h#L15